firebase-firestore/src/main/java/com/google/firebase/firestore/remote/AndroidConnectivityMonitor.java

```diff
@@ -78,34 +83,63 @@ private void configureNetworkMonitoring() {
       final DefaultNetworkCallback defaultNetworkCallback = new DefaultNetworkCallback();
       connectivityManager.registerDefaultNetworkCallback(defaultNetworkCallback);
       unregisterRunnable =
-          new Runnable() {
-            @Override
-            public void run() {
-              connectivityManager.unregisterNetworkCallback(defaultNetworkCallback);
-            }
-          };
+          () -> connectivityManager.unregisterNetworkCallback(defaultNetworkCallback);
     } else {
       NetworkReceiver networkReceiver = new NetworkReceiver();
       @SuppressWarnings("deprecation")
       IntentFilter networkIntentFilter = new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION);
       context.registerReceiver(networkReceiver, networkIntentFilter);
-      unregisterRunnable =
-          new Runnable() {
-            @Override
-            public void run() {
-              context.unregisterReceiver(networkReceiver);
-            }
-          };
+      unregisterRunnable = () -> context.unregisterReceiver(networkReceiver);
     }
   }
 
   private void configureBackgroundStateListener() {
-    BackgroundDetector.getInstance().addListener(this);
+    // Manually register an ActivityLifecycleCallback. Android's BackgroundDetector only notifies
+    // when it is certain that the app transitioned from background to foreground. Instead, we
+    // want to be notified whenever there is a slight chance that this transition happened.
+    ((Application) context.getApplicationContext())
+        .registerActivityLifecycleCallbacks(
+            new Application.ActivityLifecycleCallbacks() {
+              @Nullable Activity lastActivity = null;
+
+              @Override
+              public void onActivityCreated(
+                  @NonNull Activity activity, Bundle savedInstanceState) {}
+
+              @Override
+              public void onActivityStarted(@NonNull Activity activity) {}
+
+              @Override
+              public void onActivityResumed(@NonNull Activity activity) {
+                // Only raise the foreground notification if the same activity when to the
+                // background before. This prevents notifications when an app switches between
+                // activities.
+                if (activity == lastActivity) {
```

Comment:
@samtstern Does this code make sense? I would like to detect when an app goes into the background and then back into the foreground. I assume that in this case, we will get a notification with the same activity, whereas in the normal app lifecycle we will receive this notification with a different activity.

Do you also happen to know if the memory address here will always be the same?

Reply:
I have a feeling `activity == lastActivity` is going to be too simple, as Android destroys activities all the time. That's why you need to do all the `savedInstanceState` stuff to handle device rotation.

But I don't know the more robust way to handle `Activity` comparison.

Reply:
Moreover, holding a strong reference to an `Activity` leaks the `Activity` instance, causing a memory leak. As Sam mentions, the `Activity` instances would compare unequal even if it was the same activity but some "configuration change" caused the `Activity` to be destroyed and re-created, such as a screen rotation.

What if you instead compared the activity's `Intent` for equality? Holding a strong reference to an `Intent` is not a memory leak and it more-or-less describes the `Activity`. It's not perfect, because, in theory, you can start a new `Activity` with the exact same `Intent` overtop of itself. But I think that situation is fairly rare.

Reply:
I changed the implementation to also hook into the component callbacks, which offer a notification when an app's UI becomes hidden.  This is also used here: cs/piper///depot/google3/java/com/google/android/gmscore/integ/client/common/src/com/google/android/gms/common/api/internal/BackgroundDetector.java;l=140

The code is now very similar to BackgroundDetector, but it also fires if the state transition is not 100% certain. In my testing, it works to detect the transitions, whereas BackgroundDetector sometimes misses a transition. It is likely better to fire more often then to miss a transition.